static/src/javascripts/projects/common/modules/commercial/adfree/maintain-adfree-freshness.js

```diff
@@ -0,0 +1,36 @@
+define([
+    'common/utils/cookies',
+    'common/utils/config',
+    'common/utils/storage',
+    'common/modules/identity/api',
+    'common/modules/commercial/adfree/renew-adfree-status'
+], function (
+    cookies,
+    config,
+    storage,
+    identity,
+    renewAdfreeStatus
+) {
+    return maintainAdfreeFreshness;
+
+    function maintainAdfreeFreshness() {
+        if (featureEnabled() && identity.isUserLoggedIn() && userNeedsNewAdfreeCookie()) {
+            renewAdfreeStatus.renew();
+        }
+    }
+
+    function userNeedsNewAdfreeCookie() {
+        var adfreeCookie = cookies.get('gu_adfree_user');
+        return (adfreeCookie === null) || adfreeCookieStale();
+    }
+
+    function adfreeCookieStale() {
+        var adfreeCookieExpiry = storage.local.get('gu.adfree.user.expiry'),
```

Comment:
I do like the idea of doing this purely in localStorage - because then nextgen code is entirely responsible for reasoning about the lifecycle of this client data, rather than relying on membership-service knowing how we want to persist it - but there is a challenge to this approach: when the user logs out, how will we drop the adfree data? With a cookie, we can just have the controller return a pragma to drop it (like we do with the GU_U and SC_GU_U cookies).

We could intercept button / link events and delete the entry before redirecting the user to the signout controller, but that seems it could be porous and easy to accidentally break.

We could return a cookie that tells the client to refresh non-cookie user data, I suppose.